codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -375,13 +381,51 @@ public void write(int allowedBytes) {
                         writePromise = ctx.newPromise();
                         writePromise.addListener(this);
                     }
+                    if (toWrite instanceof SlicedByteBuf && data instanceof CompositeByteBuf) {
+                        // Free the partials as they are written
+                        final CompositeByteBuf toFree = (CompositeByteBuf) data;
+                        writePromise.addListener(new ChannelFutureListener() {
+                            @Override
+                            public void operationComplete(ChannelFuture future) throws Exception {
+                                toFree.discardReadComponents();
+                            }
+                        });
+                    }
                     frameWriter().writeData(ctx, stream.id(), toWrite, writeablePadding,
-                        size == bytesWritten && endOfStream, writePromise);
+                            size == bytesWritten && endOfStream, writePromise);
                 } while (size != bytesWritten && allowedBytes > bytesWritten);
             } finally {
                 size -= bytesWritten;
             }
         }
+
+        @Override
+        public boolean merge(Http2RemoteFlowController.FlowControlled next) {
+            // Disucssion point: If the buffer were two adjacent slices of the same underlying buffer it would
```

Comment:
I wouldn't expect this to ever happen.  When the application writes, it will likely write an entire buffer ... not sure there would ever be a reason for it to chunk.